src/freertos_drivers/common/MCP2515Can.hxx

```diff
@@ -584,7 +594,141 @@ private:
      */
     MCP2515Can();
 
+    /** Allow access to MCP2515Can from MCP2515GPO */
+    friend class MCP2515GPO;
+
+    /** Allow access to MCP2515Can from MCP2515GPI */
+    friend class MCP2515GPI;
+
     DISALLOW_COPY_AND_ASSIGN(MCP2515Can);
 };
 
+/** General Purpose Output (GPO) instance on the MCP2515.
+ */
+class MCP2515GPO : public Gpio
+{
+public:
+    /** Constructor.
+     * @param instance parrent MCP2515Can instance that "owns" the interface.
+     * @param bit bit index (0 through 1) of the output.
+     */
+    MCP2515GPO(MCP2515Can *instance, uint8_t bit)
+        : Gpio()
+        , instance_(instance)
+        , bit_(bit)
+    {
+        HASSERT(bit < 2);
+    }
+
+    /** Writes a GPO pin (set or clear to a specific state).
+     * @param new_state the desired output state.  See @ref Value.
+     */
+    void write(Value new_state) const override
+    {
+        new_state ? set() : clr();
+    }
+
+    /** Retrieves the current @ref Value of a GPO output sate (requested).
+     * @return @ref SET if currently high, @ref CLR if currently low
+     */
+    Value read() const override
+    {
+        return instance_->gpoData & (0x1 << bit_) ? Gpio::SET : Gpio::CLR;
+    }
+
+    /** Sets the GPO pin to high.
+     */
+    void set() const override
+    {
+        portENTER_CRITICAL();
+        instance_->gpoData |= 0x1 << bit_;
+        portEXIT_CRITICAL();
+        instance_->sem.post();
```

Comment:
I have another question. Can we prevent multiple posting of the semaphore by GPIO drivers?
The thread will spin so long as the semaphore is at nonzero, bu after each GPIO update we need to do only one spin.

Reply:
we could do this with a dirty bit, for example: when the thread completes the spin, it clears the bit, and we only post the semaphore if the bit is clear.

Reply:
Fixed